### ISSUE TITLE
raft: What prevents a restarting node from replaying CommittedEntries

### ISSUE THREAD

**Issue by kchristidis** [coreos/etcd#9809]
Consider the case where a Raft node crashes halfway through (or even right after) the processing of `CommittedEntries` from a `node.Ready` message.

When that node restarts what prevents it from re-applying the `CommittedEntries` that it had already applied (before crashing) in its application?

Remember that we did not have a chance to save a snapshot that includes this applied index to the WAL yet.

**Comment by lishuai87** [comment]
recover to last snapshot
https://github.com/coreos/etcd/blob/master/etcdserver/server.go#L400

**Comment by kchristidis** [comment]
@lishuai87: As I noted above:

> Remember that we did not have a chance to save a snapshot that includes this applied index to the WAL yet.

So recovering to the last snapshot alone should _not be enough_ to prevent replaying.

**Comment by kchristidis** [comment]
My take on the right answer:

Whenever you process a committed entry, the application is expected to record to permanent storage the corresponding applied index. Upon restarting then, the application restores this index to memory, and when presented with the last snapshot, it does not replay entries before that index.

TL;DR — make sure your application stores the most recently committed entry index.

Would like to get confirmation from a domain expert however.
